gno.land/pkg/gnoland/types.go

```diff
@@ -19,7 +19,92 @@ func ProtoGnoAccount() std.Account {
 	return &GnoAccount{}
 }
 
+// GnoGenesis defines the gno genesis API,
+// adopted by differing genesis state implementations
+type GnoGenesis interface {
+	// GenesisBalances returns the genesis balances associated
+	// with the Gno genesis state
+	GenesisBalances() []Balance
+
+	// GenesisTxs returns the genesis transactions associated
+	// with the Gno genesis state
+	GenesisTxs() []GenesisTx
+}
+
+type GenesisTx interface {
+	// Tx returns the standard TM2 transaction
+	Tx() std.Tx
+
+	// Metadata returns the metadata tied
+	// to the tx, if any
+	Metadata() *GenesisTxMetadata
+}
+
 type GnoGenesisState struct {
 	Balances []Balance `json:"balances"`
 	Txs      []std.Tx  `json:"txs"`
 }
+
+func (g GnoGenesisState) GenesisBalances() []Balance {
+	return g.Balances
+}
+
+func (g GnoGenesisState) GenesisTxs() []GenesisTx {
+	genesisTxs := make([]GenesisTx, len(g.Txs))
+
+	for i, tx := range g.Txs {
+		genesisTxs[i] = gnoGenesisTx{
+			tx: tx,
+		}
+	}
+
+	return genesisTxs
+}
+
+type gnoGenesisTx struct {
+	tx std.Tx
+}
+
+func (g gnoGenesisTx) Tx() std.Tx {
+	return g.tx
+}
+
+func (g gnoGenesisTx) Metadata() *GenesisTxMetadata {
+	return nil
+}
+
+type MetadataGenesisState struct {
+	Balances []Balance    `json:"balances"`
+	Txs      []MetadataTx `json:"txs"`
+}
+
+type MetadataTx struct {
+	GenesisTx  std.Tx            `json:"tx"`
+	TxMetadata GenesisTxMetadata `json:"metadata"`
```

Comment:
```suggestion
	ExecContext ExecContext `json:"exec_context"`
```

What if we reuse the `exec_context` and set the fields we want today? For example, we could start with the Timestamp and add other fields tomorrow.

Reply:
There is very little configurability of the context in relation to the VM execution context, so I think it makes sense to keep the `GenesisTxMetadata` (custom)

```go
func (c Context) WithContext(ctx context.Context) Context {
	c.ctx = ctx
	return c
}

func (c Context) WithMode(mode RunTxMode) Context {
	c.mode = mode
	return c
}

func (c Context) WithMultiStore(ms store.MultiStore) Context {
	c.ms = ms
	return c
}

func (c Context) WithBlockHeader(header abci.Header) Context {
	c.header = header
	return c
}

func (c Context) WithChainID(chainID string) Context {
	c.chainID = chainID
	return c
}

func (c Context) WithTxBytes(txBytes []byte) Context {
	c.txBytes = txBytes
	return c
}

func (c Context) WithLogger(logger *slog.Logger) Context {
	c.logger = logger
	return c
}

func (c Context) WithVoteInfos(voteInfo []abci.VoteInfo) Context {
	c.voteInfo = voteInfo
	return c
}

func (c Context) WithGasMeter(meter store.GasMeter) Context {
	c.gasMeter = meter
	return c
}

func (c Context) WithBlockGasMeter(meter store.GasMeter) Context {
	c.blockGasMeter = meter
	return c
}

func (c Context) WithMinGasPrices(gasPrices []GasPrice) Context {
	c.minGasPrices = gasPrices
	return c
}

func (c Context) WithConsensusParams(params *abci.ConsensusParams) Context {
	c.consParams = params
	return c
}

func (c Context) WithEventLogger(em *EventLogger) Context {
	c.eventLogger = em
	return c
}

// WithValue is shorthand for:
//
//	c.WithContext(context.WithValue(c.Context(), key, value))
//
// It adds a value to the [context.Context].
func (c Context) WithValue(key, value interface{}) Context {
	c.ctx = context.WithValue(c.ctx, key, value)
	return c
}

```